### PR TITLE
Revert AD "sticky gate" 

### PR DESCRIPTION
**Pull request by dgenr8**
Commit 48b712e created a "sticky gate" which, when triggered by an AD event, allows any size block for about one full day. The gate is open while an excessive block as recent as AD + 144 blocks is in the active chain.

This behavior can adversely affect consensus and convergence, and should be reverted until it can be weighed more carefully.  It should be reverted, rather than merely disabled, because not just the numerical values but the behavior itself needs to be debated.

Risk is introduced by the combination of published AD values and the daylong period devoid of blocksize checks, which present attacker a way to disrupt the BU chain.  His challenge is to create an excessive block, followed by AD confirming blocks, before honest BU nodes can create a chain of length AD+1.

If an attacker can succeed in this, he has 144 blocks which are permitted to be any size at all. He then submits as many enormous blocks as possible, to make them a permanent part of the BU chain, straining all BU users, possibly fragmenting the chain and/or pushing some users off the network.

Attacker's likelihood of success during a single attempt to open the sticky gate depends on the hashrate fraction q (of total BU hashrate) that he dedicates to this purpose. The probability of success for AD=5 is as follows (review of this is very welcome):
```
AD=5
q       P[success]
0.05	0.00000580135
0.10	0.000295706
0.15	0.00265686
0.20	0.0116542
0.25	0.0343275
0.30	0.0782248
0.35	0.148684
0.40	0.246502
0.45	0.366877
0.50	0.5
0.55	0.633123
0.60	0.753498
```
Attacker is free to make another attempt, as soon as one attempt fails. 

Two failed assertions in CheckBlockIndex were masked by the sticky gate.  To satisfy the setBlockIndexCandidates assertion, headers without data must not count toward satisfaction of AD, since CheckBlockIndex is called before data arrrives.  The fix is in isChainExcessive. To satisfy the mapBlocksUnlinked assertion, an additional status value BLOCK_EXCESSIVE_CHILD is added to children of an excessive block that are added to the map when a candidate is disqualified. The status is cleared when the blocks are moved out of mapBlocksUnlinked.

### CODE REVIEW

**Comment by gandrewstone** [comment]
If your EB/AD is exceeded, and miners set their AD rationally, a majority of nodes must have a higher EB that has not been exceeded.  These nodes will reject the huge block.   Additionally, blocks full of previously unrelayed tx propagate and validate very slowly compared to shorter blocks full of relayed validated tx.

I hear a lot of what if but no concrete analysis in this PR.

The BUIP process is the correct way to open discussion when you disagree with the developer about a commit.

**Comment by ftrader** [comment]
Are you confident that the 'BLOCK_EXCESSIVE_CHILD' condition is being tested by the modified excessive.py? I struggled to identify a relevant section due to lack of comments.

**Comment by dgenr8** [comment]
@gandrewstone The BU specifications (BUIPs) say that an excessive block will be accepted when confirmed by AD blocks.  This feature allows the network to converge.  Nowhere in the specifications is it said that this crucial feature is completely dropped for 144 blocks following a single activation.  I know it came as a surprise to me.

In this situation, I'd expect the process to be to revert the undocumented behavior quickly, and if desired, re-introduce it as a BUIP.

The risk of the sticky gate is very high, and the only apparent benefit -- that things are easier for the miner or node operator when his EB is exceeded -- is dubious.  If we want this to be easy, why have EB at all?

**Comment by dgenr8** [comment]
@ftrader If you remove the BLOCK_EXCESSIVE_CHILD logic, you will see excessive.py generate failed assertions in CheckBlockIndex.

In core, blocks with fully connected transactions are asserted not to appear in mapBlocksUnlinked (a holding area for blocks, originally just those with missing data, that might later become part of the main chain).  BU uses this same mechanism for excessive blocks, which I think is a fine solution.

Pre-48b712e, BU relaxed the connected-transactions assertion for blocks whose chain (defined as the chain leading up to that block) is considered excessive, but that's not enough. During the search for the best chain, all blocks in a new fork are checked individually, and when an excessive block is found, all of its children are placed in mapBlocksUnlinked, including the ones that are beyond an AD event and whose backing-chain is therefore not excessive.  These blocks need to be marked with a new marker BLOCK_EXCESSIVE_CHILD so they can be excluded from the connected-transactions assertion.

This parallels a design used by core, where a BLOCK_FAILED_CHILD bit is used.

48b712e addressed the problem differently, by suspending the assertion for any chain containing an excessive block in the last 144+AD blocks.

**Comment by ftrader** [comment]
Thanks @dgenr8 , also for the detailed explanation about BLOCK_EXCESSIVE_CHILD.

However, I should not have to resort to removing code from the commit to see if the test script tests a feature which you introduce - that should be visible by inspection of the script :-)

**Comment by gandrewstone** [comment]
I don't see this issue as timely which is why I am not thinking of reverting it now.  I mean we are at 10-15% hash power, we'd need to have the majority, and have a hard fork, and have an expensive attack before this logic comes into play.

The Developer is allowed to make changes as needed without a BUIP to fix issues that come up (and members are allowed to disagree by opening a BUIP), please read the Articles for more details.  There is no other way to form "consensus" as to what changes are BUIP-needed and which are not... I talked more about this on slack.

Let's talk here about solutions to this problem, not possible issues with the Bitcoin Unlimited process.

The problem this solves is the "always trailing" issue.  Basically, let's say EB is 1MB and the fork issues constant 1.1MB blocks.  In this case the node will always be AD blocks behind.  This would be very bad for miners, not great for full nodes, and is not in the spirit of the EB/AD algorithm, which is to create a barrier but move to the tip of the main chain when the barrier is broken.

When implementing this, I had another idea that solves your concern.  I originally didn't implement it because I thought it would be hard/awkward in a few ways, but now I'm thinking we can make it clean like this:

Basically, the "effective" EB (let's call that eEB) is an internal variable.  It is the max(EB, max block size in the last 144).  The node marks a block "excessive" if it exceeds the eEB.

So this means that the barrier is never broken, it is just pushed back to the largest block.  One issue with this is constant and slowly increasing block sizes.  It may make sense to use:
max(EB, (100k + max block size in the last 144))

just to keep the EB barrier from being activated over and over.  But I'd rather not have yet-another-configurable for this, I think a constant would be ok...

**Comment by dgenr8** [comment]
"Attackers" in this scenario is not limited to those who wish to harm BU. Stickiness changes the calculus of cooperating friendly miners, too.

When planning a blocksize increase, the numbers above clearly show that less than 50% is required to increase the blocksize.  40% hashrate could do it in 3 attempts with 57% success.

Does anybody really want that?

**Comment by ftrader** [comment]
>I don't see this issue as timely which is why I am not thinking of reverting it now. I mean we are at 10-15% hash power, we'd need to have the majority, and have a hard fork, and have an expensive attack before this logic comes into play.

@gandrewstone : It's as if you don't expect this to happen anytime soon.

If you use this criticism, you ought to give an indication of at what percentage of BU hashpower you would consider this problem "timely", and what you think the time window between that and "have a hard fork" would be...

**Comment by dgenr8** [comment]
[BUIP038](https://bitco.in/forum/threads/buip038-revert-sticky-gate.1624/), sponsored by @ftrader, has been written to document this proposed change.

**Comment by gandrewstone** [comment]
BUIP038 did not pass, closing this PR.  Thanks for the thought and hard work that went into identifying and characterizing this potential attack...
